### PR TITLE
Added -no-sanitize=unsigned-integer-overflow build flag

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
Changelog category (leave one):

- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
@alexey-milovidov You told that we don't enable this check in sanitizer, but seems like it is a part of `-fsanitize=integer` flag which is a part of `-fsanitize=undefined`. As it written on a random page in the internet: https://issueexplorer.com/issue/google/sanitizers/1421

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Review by KochetovNicolai — APPROVED** [review]

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
@nikitamikhaylov Unsigned integer overflow is defined in C++ and is not warned by undefined behaviour sanitizer by default.
Example:

Unsinged:
```
milovidov@milovidov-desktop:~$ clang++ -fsanitize=undefined -xc++ -include iostream - <<< "int main() { unsigned x = 4000000000; std::cerr << x * 2 << '\n'; return 0; }" && ./a.out 
3705032704
```

Signed:
```
milovidov@milovidov-desktop:~$ clang++ -fsanitize=undefined -xc++ -include iostream - <<< "int main() { int x = 2000000000; std::cerr << x * 2 << '\n'; return 0; }" && ./a.out 
<stdin>:1:49: runtime error: signed integer overflow: 2000000000 * 2 cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior <stdin>:1:49 in 
-294967296
```

File: cmake/sanitize.cmake [hunk 1]
```diff
@@ -77,7 +77,7 @@ if (SANITIZE)
         endif()
 
     elseif (SANITIZE STREQUAL "undefined")
-        set (UBSAN_FLAGS "-fsanitize=undefined -fno-sanitize-recover=all -fno-sanitize=float-divide-by-zero")
+        set (UBSAN_FLAGS "-fsanitize=undefined -fno-sanitize-recover=all -fno-sanitize=float-divide-by-zero -fno-sanitize=unsigned-integer-overflow")
```
Comment (alexey-milovidov) [inline cmake/sanitize.cmake:80]: Need to provide comments, like
- it is set implicitly by OSSFuzz (?);
- it is not undefined behaviour according to C++ standard, compilers or CPU;
- it is actively used in hash functions and random number generators.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by nikitamikhaylov** [comment]
@alexey-milovidov I want to keep my changes in master if possible, because with them fuzzers seems to work, otherwise - not. I don’t fully understand the reasons..

**Comment by alexey-milovidov** [comment]
Ok, just add the comment and merge.

**Comment by alexey-milovidov** [comment]
Maybe OSS-Fuzz is modifying the default set of flags for UBSan.

**Comment by nikitamikhaylov** [comment]
> Maybe OSS-Fuzz is modifying the default set of flags for UBSan.

Yes they do. But in a strange way. In docker image for builds they set env variable https://github.com/google/oss-fuzz/blob/fd1b3e6dbfaead8d368de51d1933ea4db91222c3/infra/base-images/base-builder/Dockerfile#L77

**Comment by nikitamikhaylov** [comment]
> Maybe OSS-Fuzz is modifying the default set of flags for UBSan.

Yes they do, but in a strange way. In docker image for build they set env variable 
https://github.com/google/oss-fuzz/blob/fd1b3e6dbfaead8d368de51d1933ea4db91222c3/infra/base-images/base-builder/Dockerfile#L77 which contains `unsigned-integer-overflow` flag. I don't really know whether compiler can read this variable during build. 

Also they have flag `silence_unsigned_overflow=1` inside the image where fuzzers are run. 
https://github.com/google/oss-fuzz/blob/fd1b3e6dbfaead8d368de51d1933ea4db91222c3/infra/base-images/base-runner/Dockerfile#L67

You may think that everything have to be Ok, but in fact it doesn't. Even if I pass `-fno-sanitize=unsigned-integer-overflow` directly to cmake during fuzzer building It doesn't help. The only way to stop failing on UIO is to set the variable in our CMakeFiles.

**Pull request merged**